File: src/lib/smp.zig
const std = @import("std");

const Emu  = @import("core/emu.zig").Emu;
const SSMP = @import("core/s_smp.zig").SSMP;

const main = @import("main.zig");
const emu  = @import("emu.zig");

pub const MemoryPage = extern struct {
    is_error: bool,
    array: [256]u8
};

pub const State = extern struct {
    timer_0_stage_0: ?[*]u8 = null,
    timer_0_stage_1: ?[*]u8 = null,
    timer_0_stage_2: ?[*]u8 = null,

    timer_1_stage_0: ?[*]u8 = null,
    timer_1_stage_1: ?[*]u8 = null,
    timer_1_stage_2: ?[*]u8 = null,

    timer_2_stage_0: ?[*]u8 = null,
    timer_2_stage_1: ?[*]u8 = null,
    timer_2_stage_2: ?[*]u8 = null,

    global_timer_disable: ?[*]u8 = null,
    ram_write_enable:     ?[*]u8 = null,
    ram_disable:          ?[*]u8 = null,
    global_timer_enable:  ?[*]u8 = null,
    ram_waitstates:       ?[*]u8 = null,
    io_waitstates:        ?[*]u8 = null,

    timer_on_flags: ?[*]u8 = null,
    use_boot_rom:   ?[*]u8 = null,

    dsp_address: ?[*]u8 = null,

    input_ports:  ?[*]u8 = null,
    output_ports: ?[*]u8 = null,

    aux: ?[*]u8 = null,

    timer_dividers: ?[*]u8 = null,
    timer_outputs:  ?[*]u8 = null
};

pub const LogSlice = extern struct {
    ptr: ?[*]Log = null,
    size: u32 = 0
};

pub const Log = extern struct {
    type: u32 = 0,

    dsp_cycle:  u64,
    address:    u16,
    
    pre_data:   u8 = 0,
    write_data: u8 = 0,
    post_data:  u8 = 0,
};

pub inline fn read_byte(address: u16, emu_ptr: ?*Emu) !u8 {
    var ep = emu.get_ptr(emu_ptr);
    try main.validate_ptr(Emu, ep);

    return ep.?.s_smp.read_data(address);
}

pub inline fn read_word(address: u16, emu_ptr: ?*Emu) !u16 {
    var ep = emu.get_ptr(emu_ptr);
    try main.validate_ptr(Emu, ep);

    const lo = ep.?.s_smp.read_data(address);
    const hi = ep.?.s_smp.read_data(address +% 1);

    return @as(u16, lo) | @as(u16, hi) << 8;
}

pub inline fn read_page(address: u16, emu_ptr: ?*Emu) !MemoryPage {
    var ep = emu.get_ptr(emu_ptr);
    try main.validate_ptr(Emu, ep);

    var page = MemoryPage {
        .is_error = false,
        .array = undefined
    };

    for (0..256) |i| {
        const ii: u16 = @intCast(i);
        page.array[ii] = ep.?.s_smp.read_data(address +% ii);
    }

    return page;
}

pub inline fn debug_read_byte(address: u16, emu_ptr: ?*Emu) !u8 {
    var ep = emu.get_ptr(emu_ptr);
    try main.validate_ptr(Emu, ep);

    return ep.?.s_smp.debug_read_data(address);
}

pub inline fn debug_read_word(address: u16, emu_ptr: ?*Emu) !u16 {
    var ep = emu.get_ptr(emu_ptr);
    try main.validate_ptr(Emu, ep);

    const lo = ep.?.s_smp.debug_read_data(address);
    const hi = ep.?.s_smp.debug_read_data(address +% 1);

    return @as(u16, lo) | @as(u16, hi) << 8;
}

pub inline fn debug_read_page(address: u16, emu_ptr: ?*Emu) !MemoryPage {
    var ep = emu.get_ptr(emu_ptr);
    try main.validate_ptr(Emu, ep);

    var page = MemoryPage {
        .is_error = false,
        .array = undefined
    };

    for (0..256) |i| {
        const ii: u16 = @intCast(i);
        page.array[ii] = ep.?.s_smp.debug_read_data(address +% ii);
    }

    return page;
}

pub inline fn write_byte(address: u16, value: u8, emu_ptr: ?*Emu) !void {
    var ep = emu.get_ptr(emu_ptr);
    try main.validate_ptr(Emu, ep);

    ep.?.s_smp.write_data(address, value);
}

pub inline fn write_word(address: u16, value: u16, emu_ptr: ?*Emu) !void {
    var ep = emu.get_ptr(emu_ptr);
    try main.validate_ptr(Emu, ep);

    ep.?.s_smp.write_data(address,      @intCast(value & 0xFF));
    ep.?.s_smp.write_data(address +% 1, @intCast(value >>   8));
}

pub inline fn get_boot_rom_ptr(emu_ptr: ?*Emu) !?[]const u8 {
    const ep = emu.get_ptr(emu_ptr);
    try main.validate_ptr(Emu, ep);

    return ep.?.s_smp.boot_rom;
}

pub inline fn get_state(emu_ptr: ?*Emu) !State {
    var ep = emu.get_ptr(emu_ptr);
    try main.validate_ptr(Emu, ep);

    const state = &ep.?.s_smp.state;

    return .{
        .timer_0_stage_0 = @ptrCast(&state.timer_states[0].stage_0),
        .timer_0_stage_1 = @ptrCast(&state.timer_states[0].stage_1),
        .timer_0_stage_2 = @ptrCast(&state.timer_states[0].stage_2),

        .timer_1_stage_0 = @ptrCast(&state.timer_states[1].stage_0),
        .timer_1_stage_1 = @ptrCast(&state.timer_states[1].stage_1),
        .timer_1_stage_2 = @ptrCast(&state.timer_states[1].stage_2),

        .timer_2_stage_0 = @ptrCast(&state.timer_states[2].stage_0),
        .timer_2_stage_1 = @ptrCast(&state.timer_states[2].stage_1),
        .timer_2_stage_2 = @ptrCast(&state.timer_states[2].stage_2),

        .global_timer_disable = @ptrCast(&state.global_timer_disable),
        .ram_write_enable     = @ptrCast(&state.ram_write_enable),
        .ram_disable          = @ptrCast(&state.ram_disable),
        .global_timer_enable  = @ptrCast(&state.global_timer_enable),
        .ram_waitstates       = @ptrCast(&state.ram_waitstates),
        .io_waitstates        = @ptrCast(&state.io_waitstates),

        .timer_on_flags = @ptrCast(&state.timer_on_flags[0]),
        .use_boot_rom   = @ptrCast(&state.use_boot_rom),

        .dsp_address = @ptrCast(&state.dsp_address),

        .input_ports  = @ptrCast(&state .input_ports[0]),
        .output_ports = @ptrCast(&state.output_ports[0]),

        .aux = @ptrCast(&state.aux[0]),

        .timer_dividers = @ptrCast(&state.timer_dividers[0]),
        .timer_outputs  = @ptrCast(&state .timer_outputs[0])
    };
}

pub inline fn enable_logging(emu_ptr: ?*Emu) !void {
    var ep = emu.get_ptr(emu_ptr);
    try main.validate_ptr(Emu, ep);

    ep.?.s_smp.enable_access_logs = true;
}

pub inline fn disable_logging(emu_ptr: ?*Emu) !void {
    var ep = emu.get_ptr(emu_ptr);
    try main.validate_ptr(Emu, ep);

    ep.?.s_smp.enable_access_logs = false;
}

pub inline fn get_access_logs(start_cycle: u64, emu_ptr: ?*Emu) !LogSlice {
    var ep = emu.get_ptr(emu_ptr);
    try main.validate_ptr(Emu, ep);

    const logs_ = ep.?.s_smp.get_access_logs_range(start_cycle);

    var logs = logs_;
    var size: u32 = 0;

    while (logs.step()) {
        size += 1;
    }

    const array = if (size > 0)
        try main.alloc.alloc(Log, size)
    else
        try main.alloc.alloc(Log, 1);

    logs = logs_;
    var i: u32 = 0;

    while (logs.step()) {
        const log = logs.value();

        array[i] = .{
            .type = switch (log.type) {
                SSMP.AccessType.none       => @intFromEnum(main.LogType.none),
                SSMP.AccessType.read       => @intFromEnum(main.LogType.read),
                SSMP.AccessType.write      => @intFromEnum(main.LogType.write),
                SSMP.AccessType.exec       => @intFromEnum(main.LogType.exec),
                SSMP.AccessType.fetch      => @intFromEnum(main.LogType.fetch),
                SSMP.AccessType.dummy_read => @intFromEnum(main.LogType.dummy_read),
            },
            .dsp_cycle  = log.dsp_cycle,
            .address    = log.address,
            .pre_data   = log.pre_data   orelse 0,
            .write_data = log.write_data orelse 0,
            .post_data  = log.post_data  orelse 0,
        };

        i += 1;
    }

    return .{
        .ptr  = @ptrCast(array.ptr),
        .size = size,
    };
}

pub inline fn free_logs(log_ptr: ?*Log) !void {
    try main.validate_ptr(Log, log_ptr);
    main.alloc.destroy(log_ptr.?);
}

File: src/cli/play/Apollo/SMP.cs
using Jimbl;

namespace Apollo;

using System.Collections;

public class SMP {
	public unsafe class Properties {
		public class TimerProps {
			public byte Stage0 {
				get {
					emu.MaybeAcquireLock();
					try {
						if      (index == 0) return *((byte*) state.Timer0Stage0);
						else if (index == 1) return *((byte*) state.Timer1Stage0);
						else                 return *((byte*) state.Timer2Stage0);
					}
					finally { emu.MaybeReleaseLock(); }
				}
				set {
					emu.MaybeAcquireLock();
					try {
						if      (index == 0) *((byte*) state.Timer0Stage0) = value;
						else if (index == 1) *((byte*) state.Timer1Stage0) = value;
						else                 *((byte*) state.Timer2Stage0) = value;
					}
					finally { emu.MaybeReleaseLock(); }
				}
			}
			
			public byte Stage1 {
				get {
					emu.MaybeAcquireLock();
					try {
						if      (index == 0) return (byte) (*((byte*) state.Timer0Stage1) & 1);
						else if (index == 1) return (byte) (*((byte*) state.Timer1Stage1) & 1);
						else                 return (byte) (*((byte*) state.Timer2Stage1) & 1);
					}
					finally { emu.MaybeReleaseLock(); }
				}
				set {
					emu.MaybeAcquireLock();
					try {
						if      (index == 0) *((byte*) state.Timer0Stage1) = (byte) (value & 1);
						else if (index == 1) *((byte*) state.Timer1Stage1) = (byte) (value & 1);
						else                 *((byte*) state.Timer2Stage1) = (byte) (value & 1);
					}
					finally { emu.MaybeReleaseLock(); }
				}
			}
			
			public byte Stage2 {
				get {
					emu.MaybeAcquireLock();
					try {
						if      (index == 0) return *((byte*) state.Timer0Stage2);
						else if (index == 1) return *((byte*) state.Timer1Stage2);
						else                 return *((byte*) state.Timer2Stage2);
					}
					finally { emu.MaybeReleaseLock(); }
				}
				set {
					emu.MaybeAcquireLock();
					try {
						if      (index == 0) *((byte*) state.Timer0Stage2) = value;
						else if (index == 1) *((byte*) state.Timer1Stage2) = value;
						else                 *((byte*) state.Timer2Stage2) = value;
					}
					finally { emu.MaybeReleaseLock(); }
				}
			}
			
			public byte Stage3 {
				get {
					emu.MaybeAcquireLock();
					try     { return Output; }
					finally { emu.MaybeReleaseLock(); }
				}
				set {
					emu.MaybeAcquireLock();
					try     { Output = value; }
					finally { emu.MaybeReleaseLock(); }
				}
			}
			
			public bool Enabled {
				get {
					emu.MaybeAcquireLock();
					try     { return (*((byte*) (state.TimerOnFlags) + index % 3) & 1) != 0; }
					finally { emu.MaybeReleaseLock(); }
				}
				set {
					emu.MaybeAcquireLock();
					try     { *((byte*) (state.TimerOnFlags) + index % 3) = (byte) (value ? 1 : 0); }
					finally { emu.MaybeReleaseLock(); }
				}
			}
			
			public byte Divider {
				get {
					emu.MaybeAcquireLock();
					try     { return *((byte*) (state.TimerDividers) + index % 3); }
					finally { emu.MaybeReleaseLock(); }
				}
				set {
					emu.MaybeAcquireLock();
					try     { *((byte*) (state.TimerDividers) + index % 3) = value; }
					finally { emu.MaybeReleaseLock(); }
				}
			}
			
			public byte Output {
				get {
					emu.MaybeAcquireLock();
					try     { return (byte) (*((byte*) (state.TimerOutputs) + index % 3) & 0xF); }
					finally { emu.MaybeReleaseLock(); }
				}
				set {
					emu.MaybeAcquireLock();
					try     { *((byte*) (state.TimerOutputs) + index % 3) = (byte) (value & 0xF); }
					finally { emu.MaybeReleaseLock(); }
				}
			}
			
			Emulator     emu;
			DLL.SmpState state;
			int          index;
		
			internal TimerProps(Emulator emu, DLL.SmpState state, int index) {
				this.emu   = emu;
				this.state = state;
				this.index = index;
			}
		}
		
		public class APUIO {
			public class Ports: IEnumerable<byte> {
				public byte this[int index] {
					get {
						emu.MaybeAcquireLock();
						try     { return *((byte*) basePtr + (index & 3)); }
						finally { emu.MaybeReleaseLock(); }
					}
					set {
						emu.MaybeAcquireLock();
						try     { *((byte*) basePtr + (index & 3)) = value; }
						finally { emu.MaybeReleaseLock(); }
					}
				}
				
				Emulator     emu;
				DLL.SmpState state;
				IntPtr       basePtr;
		
				internal Ports(Emulator emu, DLL.SmpState state, IntPtr basePtr) {
					this.emu     = emu;
					this.state   = state;
					this.basePtr = basePtr;
				}

				public IEnumerator<byte> GetEnumerator() {
					for (var i = 0; i < 4; i++) {
						yield return this[i];
					}
				}
				
				IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
			}
			
			public Ports Input  { get; }
			public Ports Output { get; }
			
			DLL.SmpState state;
		
			internal APUIO(Emulator emu, DLL.SmpState state) {
				this.state = state;
				
				Input  = new(emu, state, state .InputPorts);
				Output = new(emu, state, state.OutputPorts);
			}
		}
		
		public TimerProps[] Timer { get; }
		public APUIO        IO    { get; }
		public UInt8Buffer  Aux   { get; }
		
		public bool GlobalTimerDisable {
			get {
				emu.MaybeAcquireLock();
				try     { return (*((byte*) state.GlobalTimerDisable) & 1) != 0; }
				finally { emu.MaybeReleaseLock(); }
			}
			set {
				emu.MaybeAcquireLock();
				try     { *((byte*) state.GlobalTimerDisable) = (byte) (value ? 1 : 0); }
				finally { emu.MaybeReleaseLock(); }
			}
		}
		
		public bool RAMWriteEnable {
			get {
				emu.MaybeAcquireLock();
				try     { return (*((byte*) state.RamWriteEnable) & 1) != 0; }
				finally { emu.MaybeReleaseLock(); }
			}
			set {
				emu.MaybeAcquireLock();
				try     { *((byte*) state.RamWriteEnable) = (byte) (value ? 1 : 0); }
				finally { emu.MaybeReleaseLock(); }
			}
		}
		
		public bool RAMDisable {
			get {
				emu.MaybeAcquireLock();
				try     { return (*((byte*) state.RamDisable) & 1) != 0; }
				finally { emu.MaybeReleaseLock(); }
			}
			set {
				emu.MaybeAcquireLock();
				try     { *((byte*) state.RamDisable) = (byte) (value ? 1 : 0); }
				finally { emu.MaybeReleaseLock(); }
			}
		}
		
		public bool GlobalTimerEnable {
			get {
				emu.MaybeAcquireLock();
				try     { return (*((byte*) state.GlobalTimerEnable) & 1) != 0; }
				finally { emu.MaybeReleaseLock(); }
			}
			set {
				emu.MaybeAcquireLock();
				try     { *((byte*) state.GlobalTimerEnable) = (byte) (value ? 1 : 0); }
				finally { emu.MaybeReleaseLock(); }
			}
		}
			
		public byte RAMWaitstates {
			get {
				emu.MaybeAcquireLock();
				try     { return (byte) (*((byte*) state.RamWaitstates) & 3); }
				finally { emu.MaybeReleaseLock(); }
			}
			set {
				emu.MaybeAcquireLock();
				try     { *((byte*) state.RamWaitstates) = (byte) (value & 3); }
				finally { emu.MaybeReleaseLock(); }
			}
		}
			
		public byte IOWaitstates {
			get {
				emu.MaybeAcquireLock();
				try     { return (byte) (*((byte*) state.IoWaitstates) & 3); }
				finally { emu.MaybeReleaseLock(); }
			}
			set {
				emu.MaybeAcquireLock();
				try     { *((byte*) state.IoWaitstates) = (byte) (value & 3); }
				finally { emu.MaybeReleaseLock(); }
			}
		}
		
		public bool UseBootROM {
			get {
				emu.MaybeAcquireLock();
				try     { return (*((byte*) state.UseBootRom) & 1) != 0; }
				finally { emu.MaybeReleaseLock(); }
			}
			set {
				emu.MaybeAcquireLock();
				try     { *((byte*) state.UseBootRom) = (byte) (value ? 1 : 0); }
				finally { emu.MaybeReleaseLock(); }
			}
		}
			
		public byte DSPAddress {
			get {
				emu.MaybeAcquireLock();
				try     { return *((byte*) state.DspAddress); }
				finally { emu.MaybeReleaseLock(); }
			}
			set {
				emu.MaybeAcquireLock();
				try     { *((byte*) state.DspAddress) = value; }
				finally { emu.MaybeReleaseLock(); }
			}
		}
			
		public byte DSPData {
			get => emu.DSP.Register[DSPAddress & 0x7F];
			set => emu.DSP.Register[DSPAddress & 0x7F] = value;
		}
		
		Emulator     emu;
		DLL.SmpState state;
		
		internal Properties(DLL.SmpState state, Emulator emu) {
			this.emu   = emu;
			this.state = state;
			
			Timer = new TimerProps[] { new(emu, state, 0), new(emu, state, 1), new(emu, state, 2) };
			IO    = new(emu, state);
			
			if (emu.MakeShared) {
				Aux = new UInt8BufferShared(emu, (byte*) state.Aux, 2);
			}
			else {
				Aux = new((byte*) state.Aux, 2);
			}
		}
	}
	
	public unsafe class Buffer {
		DLL.SmpMemoryPage page;
		
		public byte this[int index] => page.Array[index & 0xFF];
		
		internal Buffer(DLL.SmpMemoryPage page) {
			this.page = page;
		}
	}
	
	public class MemAccessLog {
		public enum LogType {
			None      = 0,
			Read      = 1,
			Write     = 2,
			Exec      = 3,
			Fetch     = 4,
			DummyRead = 5
		}
		
		public LogType Type      { get; }
		public UInt64  DSPCycle  { get; }
		public UInt16  Address   { get; }
		
		public byte?   PreData   { get; }
		public byte?   WriteData { get; }
		public byte?   PostData  { get; }
		
		internal MemAccessLog(DLL.SmpLog log) {
			Type      = (LogType) log.Type;
			DSPCycle  = log.DspCycle;
			Address   = log.Address;
			
			PreData   = Type == LogType.Write ? log.PreData   : null;
			WriteData = Type == LogType.Write ? log.WriteData : null;
			PostData  = Type == LogType.Write ? log.PostData  : null;
		}
	}
	
	bool loggingEnabled = false;
	
	public Emulator    Emulator { get; init; }
	public UInt8Buffer BootROM  { get; init; }
	
	public Properties State { get; }
	
	public bool LoggingEnabled {
		get {
			Emulator.MaybeAcquireLock();
			try     { return loggingEnabled; }
			finally { Emulator.MaybeReleaseLock(); }
		}
		set {
			Emulator.MaybeAcquireLock();
			
			try {
				bool result;
				
				if (value) {
					result = DLL.SmpEnableLogging(Emulator.handle);
				}
				else {
					result = DLL.SmpDisableLogging(Emulator.handle);
				}
				
				if (!result) {
					var errorCode = DLL.EmuGetLastError(Emulator.handle);
					Error.Throw(errorCode);
				}
				
				loggingEnabled = value;
			}
			finally { Emulator.MaybeReleaseLock(); }
		}
	}
	
	public byte ReadByte(UInt16 address) {
		Emulator.MaybeAcquireLock();
		try {
			var result = DLL.SmpReadByte(address, Emulator.handle);
			Emulator.CheckForError();
			return result;
		}
		finally { Emulator.MaybeReleaseLock(); }
	}
	
	public UInt16 ReadWord(UInt16 address) {
		Emulator.MaybeAcquireLock();
		try {
			var result = DLL.SmpReadWord(address, Emulator.handle);
			Emulator.CheckForError();
			return result;
		}
		finally { Emulator.MaybeReleaseLock(); }
	}
	
	public Buffer ReadPage(UInt16 address) {
		Emulator.MaybeAcquireLock();
		try {
			var page = DLL.SmpReadPage(address, Emulator.handle);
			if (page.IsError) {
				var errorCode = DLL.EmuGetLastError(Emulator.handle);
				Error.Throw(errorCode);
			}
			return new(page);
		}
		finally { Emulator.MaybeReleaseLock(); }
	}
	
	public byte DebugReadByte(UInt16 address) {
		Emulator.MaybeAcquireLock();
		try {
			var result = DLL.SmpDebugReadByte(address, Emulator.handle);
			Emulator.CheckForError();
			return result;
		}
		finally { Emulator.MaybeReleaseLock(); }
	}
	
	public UInt16 DebugReadWord(UInt16 address) {
		Emulator.MaybeAcquireLock();
		try {
			var result = DLL.SmpDebugReadWord(address, Emulator.handle);
			Emulator.CheckForError();
			return result;
		}
		finally { Emulator.MaybeReleaseLock(); }
	}
	
	public Buffer DebugReadPage(UInt16 address) {
		Emulator.MaybeAcquireLock();
		try {
			var page = DLL.SmpDebugReadPage(address, Emulator.handle);
			if (page.IsError) {
				var errorCode = DLL.EmuGetLastError(Emulator.handle);
				Error.Throw(errorCode);
			}
			return new(page);
		}
		finally { Emulator.MaybeReleaseLock(); }
	}
	
	public void WriteByte(UInt16 address, byte value) {
		Emulator.MaybeAcquireLock();
		try {
			var result = DLL.SmpWriteByte(address, value, Emulator.handle);
			if (!result) {
				var errorCode = DLL.EmuGetLastError(Emulator.handle);
				Error.Throw(errorCode);
			}
		}
		finally { Emulator.MaybeReleaseLock(); }
	}
	
	public void WriteWord(UInt16 address, UInt16 value) {
		Emulator.MaybeAcquireLock();
		try {
			var result = DLL.SmpWriteWord(address, value, Emulator.handle);
			if (!result) {
				var errorCode = DLL.EmuGetLastError(Emulator.handle);
				Error.Throw(errorCode);
			}
		}
		finally { Emulator.MaybeReleaseLock(); }
	}
	
	public MemAccessLog[] GetAccessLogs(Int64 startCycle) {
		Emulator.MaybeAcquireLock();
		try {
			unsafe {
				var logSlice = DLL.SmpGetAccessLogs(startCycle.SafeUnsigned(), Emulator.handle);
				if (logSlice.LogArray == IntPtr.Zero) {
					var errorCode = DLL.EmuGetLastError(Emulator.handle);
					Error.Throw(errorCode);
				}
				
				try {
					var arrayPtr = (DLL.SmpLog*) logSlice.LogArray;
					var length   = (int) logSlice.Length;
				
					var logs = new MemAccessLog[length];
					Span<DLL.SmpLog> span = new(arrayPtr, length);
				
					for (var i = 0; i < length; i++) {
						var log = span[i];
						logs[i] = new(log);
					}
					
					return logs;
				}
				finally {
					var result = DLL.SmpFreeLogs(logSlice.LogArray);
					if (!result) {
						var errorCode = DLL.GetLastError();
						Error.Throw(errorCode);
					}
				}
			}
		}
		finally { Emulator.MaybeReleaseLock(); }
	}
	
	public MemAccessLog[] GetAccessLogsDeduped(Int64 startCycle) {
		var allLogs = GetAccessLogs(startCycle);
		List<MemAccessLog> deduped = new();
		
		List<MemAccessLog> curCycleLogs = new();
		var lastCycle = startCycle.SafeUnsigned();
		
		void dedupe() {
			// Check if any read logs are present on the last cycle
			var hasReads = false;
			
			foreach (var curCycleLog in curCycleLogs) {
				if (curCycleLog.Type is MemAccessLog.LogType.Fetch or MemAccessLog.LogType.DummyRead) {
					hasReads = true;
					break;
				}
			}
			
			// No need to keep 'Read' logs if another, more specific type of read was logged on the same cycle
			if (hasReads) {
				deduped.AddRange(curCycleLogs.Where(x => x.Type != MemAccessLog.LogType.Read));
			}
			else {
				deduped.AddRange(curCycleLogs);
			}
		}
		
		foreach (var log in allLogs) {
			if (log.DSPCycle != lastCycle) {
				dedupe();
			
				lastCycle = log.DSPCycle;
				curCycleLogs.Clear();
			}
			curCycleLogs.Add(log);
		}
		dedupe();
		
		return deduped.ToArray();
	}
	
	internal SMP(Emulator emulator) {
		unsafe {
			Emulator = emulator;
			
			var bootRomPtr = DLL.SmpGetBootRomPtr(Emulator.handle);
			if (bootRomPtr == IntPtr.Zero) throw new StateError();
			
			BootROM = new((byte*) bootRomPtr, 0x40, isReadonly: true);
			
			var state = DLL.SmpGetState(Emulator.handle);
			if (state.GlobalTimerDisable == IntPtr.Zero) throw new StateError();
			
			State = new(state, emulator);
		}
	}
}